sled-agent/src/boot_disk_os_writer.rs

```diff
@@ -0,0 +1,1433 @@
+    }
+
+    async fn start_update_impl<S, Writer>(
+        &self,
+        boot_disk: M2Slot,
+        disk_devfs_path: Utf8PathBuf,
+        update_id: Uuid,
+        sha3_256_digest: [u8; 32],
+        image_upload: S,
+        disk_writer: Writer,
+    ) -> Result<(), Arc<BootDiskOsWriteError>>
+    where
+        S: Stream<Item = Result<Bytes, HttpError>> + Send + 'static,
+        Writer: DiskInterface + Send + Sync + 'static,
+    {
+        // Construct a closure that will spawn a task to drive this update, but
+        // don't actually start it yet: we only allow an update to start if
+        // there's not currently an update running targetting the same slot, so
+        // we'll spawn this after checking that below.
+        let spawn_update_task = || {
+            let (uploaded_image_tx, uploaded_image_rx) = oneshot::channel();
+            let (progress_tx, progress_rx) = watch::channel(
+                BootDiskOsWriteProgress::ReceivingUploadedImage {
+                    bytes_received: 0,
+                },
+            );
+            let (complete_tx, complete_rx) = oneshot::channel();
+            let task = BootDiskOsWriteTask {
+                log: self
+                    .log
+                    .new(slog::o!("update_id" => update_id.to_string())),
+                disk_devfs_path,
+                sha3_256_digest,
+                progress_tx,
+                disk_interface: disk_writer,
+            };
+            tokio::spawn(task.run(
+                image_upload,
+                uploaded_image_tx,
+                complete_tx,
+            ));
+            (
+                uploaded_image_rx,
+                TaskRunningState { update_id, progress_rx, complete_rx },
+            )
+        };
+
+        // Either call `spawn_update_task` and get back the handle to
+        // `uploaded_image_rx`, or return an error (if another update for this
+        // boot disk is still running).
+        let uploaded_image_rx = {
+            let mut states = self.states.lock().unwrap();
+            match states.entry(boot_disk) {
+                Entry::Vacant(slot) => {
+                    let (uploaded_image_rx, running) = spawn_update_task();
+                    slot.insert(WriterState::TaskRunning(running));
+                    uploaded_image_rx
+                }
+                Entry::Occupied(mut slot) => match slot.get_mut() {
+                    WriterState::TaskRunning(running) => {
+                        // Check whether the task is _actually_ still running,
+                        // or whether it's done and just waiting for us to
+                        // realize it.
+                        match running.complete_rx.try_recv() {
+                            Ok(_prev_result) => {
+                                // A previous write is done, but we're
+                                // immedately starting a new one, so discard the
+                                // previous result.
```

Comment:
Yeah, in hindsight since we did this for both the SP and wicketd, it seems reasonable to do it here too. Changed in [8a65c20](https://github.com/oxidecomputer/omicron/pull/4633/commits/8a65c206c03d2229270e4fc2a736f9373bcabc81).